### PR TITLE
MoME integration

### PR DESCRIPTION
**Pull request by edamamez**
Issue ref: https://github.com/lamini-ai/lamini-platform/issues/4919
05/02 demo [slide](https://docs.google.com/presentation/d/1bQ_P7gCMppwidV1HWT4whDebwH-NQFSm/edit?usp=sharing&ouid=112245441450491692733&rtpof=true&sd=true)

### MoME Adapter Feature

- [x] Static CUDA Graph
   weight tensor initialized when engine start
- [x] Multi MoME Adapters
   - [x] use mome_index to get different adapter weight to switch 
- [x] KNN mome index search replace faiss

### Develop Process

- MoME Layers:  (torch.nn.functional)
  - [x] MoMEAttentionLayer
  - [x] LoraMLPAdaptor
  - [x] LogitsProcessorWithMoME

- VLLM Adapter mannager for MoME
  - [x] MoMELayerWeights: store the weight
  - [x]  MoMEModel
  - [x]  MoMEModelManager and LRUCacheMoMEModelManager
  - [x] WorkerMoMEManager and LRUCacheWorkerMoMEManager
  - [x] Dummy Run

- VLLM Engine modify for MoME
  - [x] Config
     add mome_config config model metadate
  - [x] Request
     add MoMERequest class for all mome model process
  - [x] Scheduler
     add mome request to Sequence
  - [x] Sequence
  - [x] Worker
  - [x] ModelRunner
  - [x] Statue and Metrics

### Future Develop
- TP, PP support
- use fused op like Punica lora (https://arxiv.org/abs/2310.18547)

### CODE REVIEW

**Comment by edamamez** [comment]
Todo: update `init` functions for classes in `vllm/mome/layers.py`

**Comment by edamamez** [comment]
Building image here: `powerml/inference-engine-amd:v0.7.2-mome-test`

```
llama@coresite-sv7-106-amd-mi300-39:~/eda-test/lamini-vllm$ git pull
llama@coresite-sv7-106-amd-mi300-39:~/eda-test/lamini-vllm$ sudo docker login
llama@coresite-sv7-106-amd-mi300-39:~/eda-test/lamini-vllm$ sudo DOCKER_BUILDKIT=1 docker build -f Dockerfile.rocm -t powerml/inference-engine-amd:v0.7.2-mome-test . --build-arg max_jobs=100 --build-arg nvcc_threads=8 --build-arg RUN_WHEEL_CHECK=false

llama@coresite-sv7-106-amd-mi300-39:~/eda-test/lamini-vllm$ sudo docker push powerml/inference-engine-amd:v0.7.2-mome-test
```

and `powerml/inference-engine-nvidia:v0.7.2-mome-test`
```
llama@coresite-sv7-106-amd-mi300-39:~/eda-test/lamini-vllm$ sudo DOCKER_BUILDKIT=1 docker build . --target vllm-openai --tag powerml/inference-engine-nvidia:v0.7.2-mome-test --build-arg max_jobs=100 --build-arg nvcc_threads=8 --build-arg RUN_WHEEL_CHECK=false
```

File: vllm/mome/lamini_index.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    
+    def get_key_and_value(self, query_embeddings: torch.Tensor, k: int) -> tuple:
+        # logger.debug(f"query_embeddings shape: {query_embeddings.shape}")
+        device = query_embeddings.device
+        dtype = query_embeddings.dtype
+
+        query_norm = torch.nn.functional.normalize(query_embeddings.float(), dim=-1)
+        keys_norm = torch.nn.functional.normalize(self.keys.float(), dim=-1)
+
+        similarities = torch.matmul(query_norm, keys_norm.T)  # [B, N]
```
Comment (vincentoy) [inline vllm/mome/lamini_index.py:69]: You can use torch.cdist() here. It might be faster
`similarities = 1 - torch.cdist(query_embeddings.float(), self.keys.float(), p=2)`

File: vllm/mome/lamini_index.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            lamini_index.values = torch.from_numpy(np.load(values_path_npy)).float()
+        else:
+            raise ValueError("Values file not found")
+
+        lamini_index.keys = lamini_index.keys.to(device)
+        lamini_index.values = lamini_index.values.to(device)
+
+        return lamini_index
+    
+    def get_key_and_value(self, query_embeddings: torch.Tensor, k: int) -> tuple:
```
Comment (vincentoy) [inline vllm/mome/lamini_index.py:63]: I want to note: the logic here is not exactly the same as the original logic. The original one repeats the first item of the topK k times. It might be a mistake. So our implementation here is fine. Just want to highlight the difference

File: requirements-common.txt [hunk 3]
```diff
@@ -37,3 +37,6 @@ einops # Required for Qwen2-VL.
 compressed-tensors == 0.9.1 # required for compressed-tensors
 depyf==0.18.0 # required for profiling and debugging with compilation config
 cloudpickle # allows pickling lambda functions in model_executor/models/registry.py
+
+faiss-cpu # [Lamini] required for MoME models
```
Comment (vincentoy) [inline requirements-common.txt]: Maybe we can remove it since we are not using faiss

File: vllm/mome/lamini_index.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        if os.path.exists(values_path_json):
+            with open(values_path_json, "r") as f:
+                lamini_index.values = torch.tensor(json.load(f), dtype=torch.float32)
+        elif os.path.exists(values_path_npy):
+            lamini_index.values = torch.from_numpy(np.load(values_path_npy)).float()
+        else:
+            raise ValueError("Values file not found")
+
+        lamini_index.keys = lamini_index.keys.to(device)
+        lamini_index.values = lamini_index.values.to(device)
```
Comment (vincentoy) [inline vllm/mome/lamini_index.py:55]: Maybe we can set the keys and values to the same dtype as the model.dtype

File: vllm/mome/lamini_index.py (hunk 1 above)
Comment (haofrank) [reply vllm/mome/lamini_index.py:69]: using cosine similarity via normalized dot product, which is more appropriate for embedding comparisons than Euclidean distance. especially since 1 - cdist can produce negative values and lacks a clear similarity scale.

Comment (haofrank) [reply vllm/mome/lamini_index.py:69]: and cdist can be slower on large batches due to less optimized broadcasting and memory usage.

File: vllm/mome/lamini_index.py (hunk 4 above)
Comment (haofrank) [reply vllm/mome/lamini_index.py:55]: yes, use the same dtype with model.

File: vllm/mome/request.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (jameslamini) [inline vllm/mome/request.py]: What is this for? Is it required to be Apache-2.0 open source since this is an internal MoME related?
I notice many other files do not have this, such as 
vllm/mome/lamini_index.py
vllm/mome/mome.py
vllm/mome/models.py
vllm/mome/layers.py
etc.

Comment (haofrank) [reply vllm/mome/request.py]: removed SPDX-License-Identifier for mome related files
